File: jibjab/jjswift/Sources/jjswift/JJ/Parser.swift
/// JibJab Parser - Parses tokens into AST
/// Uses emit values from common/jj.json via JJConfig
import Foundation

public class Parser {
    private var tokens: [Token]
    private var pos: Int = 0

    public init(tokens: [Token]) {
        // Filter out newlines
        self.tokens = tokens.filter { $0.type != .newline }
    }

    private func peek(offset: Int = 0) -> Token {
        let idx = pos + offset
        if idx < tokens.count {
            return tokens[idx]
        }
        return Token(type: .eof, value: nil, line: 0, col: 0)
    }

    private func advance() -> Token {
        let token = peek()
        pos += 1
        return token
    }

    private func match(_ types: TokenType...) -> Token? {
        if types.contains(peek().type) {
            return advance()
        }
        return nil
    }

    private func expect(_ type: TokenType) throws -> Token {
        if peek().type == type {
            return advance()
        }
        let token = peek()
        let tokenText = token.value.map { "\($0)" } ?? Self.tokenSymbol(token.type)
        throw ParserError.unexpectedToken(expected: type, got: token.type, gotValue: tokenText, line: token.line)
    }

    static func tokenSymbol(_ type: TokenType) -> String {
        let key = "\(type)"
        // Check shared config first
        if let symbols = JJ.tokenSymbols, let symbol = symbols[key] {
            return symbol
        }
        // Operator symbols from config
        switch type {
        case .add: return JJ.operators.add.symbol
        case .sub: return JJ.operators.sub.symbol
        case .mul: return JJ.operators.mul.symbol
        case .div: return JJ.operators.div.symbol
        case .mod: return JJ.operators.mod.symbol
        case .eq: return JJ.operators.eq.symbol
        case .neq: return JJ.operators.neq.symbol
        case .lt: return JJ.operators.lt.symbol
        case .lte: return JJ.operators.lte.symbol
        case .gt: return JJ.operators.gt.symbol
        case .gte: return JJ.operators.gte.symbol
        case .and: return JJ.operators.and.symbol
        case .or: return JJ.operators.or.symbol
        case .not: return JJ.operators.not.symbol
        case .comma: return ","
        case .eof: return "end of file"
        case .newline: return "newline"
        case .blockEnd: return JJ.blocks.end
        default: return key
        }
    }

    public func parse() throws -> Program {
        var statements: [ASTNode] = []
        while peek().type != .eof {
            if let stmt = try parseStatement() {
                statements.append(stmt)
            } else {
                let bad = advance()
                let tokenText = bad.value.map { "\($0)" } ?? "\(bad.type)"
                throw ParserError.unrecognizedStatement(token: tokenText, line: bad.line)
            }
        }
        return Program(statements: statements)
    }

    private func parseStatement() throws -> ASTNode? {
        switch peek().type {
        case .print:
            return try parsePrint()
        case .log:
            return try parseLog()
        case .snag:
            return try parseVarDecl()
        case .loop:
            return try parseLoop()
        case .when:
            return try parseIf()
        case .morph:
            return try parseFuncDef()
        case .yeet:
            return try parseReturn()
        case .kaboom:
            return try parseThrow()
        case .enum:
            return try parseEnumDef()
        case .`try`:
            return try parseTry()
        case .comment:
            let token = advance()
            return CommentNode(text: (token.value as? String) ?? "")
        default:
            return nil
        }
    }

    private func parsePrint() throws -> PrintStmt {
        _ = advance() // PRINT
        _ = try expect(.action)
        _ = try expect(.emit)
        _ = try expect(.lparen)
        let expr = try parseExpression()
        _ = try expect(.rparen)
        return PrintStmt(expr: expr)
    }

    private func parseLog() throws -> LogStmt {
        _ = advance() // LOG
        _ = try expect(.action)
        _ = try expect(.emit)
        _ = try expect(.lparen)
        let expr = try parseExpression()
        _ = try expect(.rparen)
        return LogStmt(expr: expr)
    }

    private func parseVarDecl() throws -> VarDecl {
        _ = advance() // SNAG
        _ = try expect(.lbrace)
        let nameToken = try expect(.identifier)
        let name = (nameToken.value as? String) ?? ""
        _ = try expect(.rbrace)
        _ = try expect(.action)
        _ = try expect(.val)
        _ = try expect(.lparen)
        let value = try parseExpression()
        _ = try expect(.rparen)
        return VarDecl(name: name, value: value)
    }

    private func parseLoop() throws -> LoopStmt {
        let token = advance() // LOOP with value
        let loopSpec = (token.value as? String) ?? ""
        let body = try parseBlock()

        // Parse loop specification
        if loopSpec.contains("..") {
            let parts = loopSpec.split(separator: ":", maxSplits: 1).map { String($0) }
            let varName = parts[0]
            let rangeParts = parts[1].split(separator: "..", maxSplits: 1).map { String($0) }
            let start = try parseInlineExpr(rangeParts[0])
            let end = try parseInlineExpr(rangeParts[1])
            return LoopStmt(var: varName, start: start, end: end, collection: nil, condition: nil, body: body)
        } else if loopSpec.contains(":") {
            let parts = loopSpec.split(separator: ":", maxSplits: 1).map { String($0) }
            let varName = parts[0]
            let collection = VarRef(name: parts[1])
            return LoopStmt(var: varName, start: nil, end: nil, collection: collection, condition: nil, body: body)
        } else {
            let condition = try parseInlineExpr(loopSpec)
            return LoopStmt(var: "_", start: nil, end: nil, collection: nil, condition: condition, body: body)
        }
    }

    private func parseIf() throws -> IfStmt {
        let token = advance() // WHEN with condition
        let condition = try parseInlineExpr((token.value as? String) ?? "")
        let thenBody = try parseBlock()
        var elseBody: [ASTNode]? = nil

        if peek().type == .else {
            _ = advance()
            elseBody = try parseBlock()
        }

        return IfStmt(condition: condition, thenBody: thenBody, elseBody: elseBody)
    }

    private func parseTry() throws -> TryStmt {
        _ = advance() // consume .try
        let tryBody = try parseBlock()
        var oopsBody: [ASTNode]? = nil
        var oopsVar: String? = nil
        if peek().type == .oops {
            _ = advance()
            if peek().type == .identifier {
                oopsVar = advance().value as? String
            }
            oopsBody = try parseBlock()
        }
        return TryStmt(tryBody: tryBody, oopsBody: oopsBody, oopsVar: oopsVar)
    }

    private func parseFuncDef() throws -> FuncDef {
        let token = advance() // MORPH with signature
        let sig = (token.value as? String) ?? ""

        // Parse signature like "funcName(param1, param2)"
        guard let regex = try? NSRegularExpression(pattern: "(\\w+)\\(([^)]*)\\)") else {
            throw ParserError.invalidFunctionSignature(sig)
        }
        let range = NSRange(sig.startIndex..., in: sig)
        guard let match = regex.firstMatch(in: sig, range: range) else {
            throw ParserError.invalidFunctionSignature(sig)
        }

        guard let nameRange = Range(match.range(at: 1), in: sig),
              let paramsRange = Range(match.range(at: 2), in: sig) else {
            throw ParserError.invalidFunctionSignature(sig)
        }
        let name = String(sig[nameRange])
        let paramsStr = String(sig[paramsRange])
        let params = paramsStr.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }

        let body = try parseBlock()
        return FuncDef(name: name, params: params, body: body)
    }

    private func parseReturn() throws -> ReturnStmt {
        _ = advance() // YEET
        _ = try expect(.lbrace)
        let value = try parseExpression()
        _ = try expect(.rbrace)
        return ReturnStmt(value: value)
    }

    private func parseThrow() throws -> ThrowStmt {
        _ = advance() // KABOOM
        _ = try expect(.lbrace)
        let value = try parseExpression()
        _ = try expect(.rbrace)
        return ThrowStmt(value: value)
    }

    private func parseEnumDef() throws -> EnumDef {
        _ = advance() // ENUM
        _ = try expect(.lbrace)
        let nameToken = try expect(.identifier)
        let name = (nameToken.value as? String) ?? ""
        _ = try expect(.rbrace)
        _ = try expect(.action)
        _ = try expect(.cases)
        _ = try expect(.lparen)
        var cases: [String] = []
        if peek().type != .rparen {
            let caseToken = try expect(.identifier)
            cases.append((caseToken.value as? String) ?? "")
            while match(.comma) != nil {
                let caseToken = try expect(.identifier)
                cases.append((caseToken.value as? String) ?? "")
            }
        }
        _ = try expect(.rparen)
        return EnumDef(name: name, cases: cases)
    }

    private func parseBlock() throws -> [ASTNode] {
        var statements: [ASTNode] = []
        while peek().type != .blockEnd && peek().type != .else && peek().type != .oops && peek().type != .eof {
            if let stmt = try parseStatement() {
                statements.append(stmt)
            } else {
                let bad = advance()
                let tokenText = bad.value.map { "\($0)" } ?? "\(bad.type)"
                throw ParserError.unrecognizedStatement(token: tokenText, line: bad.line)
            }
        }
        if peek().type == .blockEnd {
            _ = advance()
        }
        return statements
    }

    private func parseExpression() throws -> ASTNode {
        return try parseOr()
    }

    private func parseOr() throws -> ASTNode {
        var left = try parseAnd()
        while match(.or) != nil {
            let right = try parseAnd()
            left = BinaryOp(left: left, op: JJ.operators.or.emit, right: right)
        }
        return left
    }

    private func parseAnd() throws -> ASTNode {
        var left = try parseEquality()
        while match(.and) != nil {
            let right = try parseEquality()
            left = BinaryOp(left: left, op: JJ.operators.and.emit, right: right)
        }
        return left
    }

    private func parseEquality() throws -> ASTNode {
        var left = try parseComparison()
        while true {
            if match(.eq) != nil {
                left = BinaryOp(left: left, op: JJ.operators.eq.emit, right: try parseComparison())
            } else if match(.neq) != nil {
                left = BinaryOp(left: left, op: JJ.operators.neq.emit, right: try parseComparison())
            } else {
                break
            }
        }
        return left
    }

    private func parseComparison() throws -> ASTNode {
        var left = try parseAdditive()
        while true {
            if match(.lte) != nil {
                left = BinaryOp(left: left, op: JJ.operators.lte.emit, right: try parseAdditive())
            } else if match(.lt) != nil {
                left = BinaryOp(left: left, op: JJ.operators.lt.emit, right: try parseAdditive())
            } else if match(.gte) != nil {
                left = BinaryOp(left: left, op: JJ.operators.gte.emit, right: try parseAdditive())
            } else if match(.gt) != nil {
                left = BinaryOp(left: left, op: JJ.operators.gt.emit, right: try parseAdditive())
            } else {
                break
            }
        }
        return left
    }

    private func parseAdditive() throws -> ASTNode {
        var left = try parseMultiplicative()
        while true {
            if match(.add) != nil {
                left = BinaryOp(left: left, op: JJ.operators.add.emit, right: try parseMultiplicative())
            } else if match(.sub) != nil {
                left = BinaryOp(left: left, op: JJ.operators.sub.emit, right: try parseMultiplicative())
            } else {
                break
            }
        }
        return left
    }

    private func parseMultiplicative() throws -> ASTNode {
        var left = try parseUnary()
        while true {
            if match(.mul) != nil {
                left = BinaryOp(left: left, op: JJ.operators.mul.emit, right: try parseUnary())
            } else if match(.div) != nil {
                left = BinaryOp(left: left, op: JJ.operators.div.emit, right: try parseUnary())
            } else if match(.mod) != nil {
                left = BinaryOp(left: left, op: JJ.operators.mod.emit, right: try parseUnary())
            } else {
                break
            }
        }
        return left
    }

    private func parseUnary() throws -> ASTNode {
        if match(.not) != nil {
            return UnaryOp(op: JJ.operators.not.emit, operand: try parseUnary())
        }
        return try parsePrimary()
    }

    private func parsePrimary() throws -> ASTNode {
        // Parentheses: either grouped expression or tuple
        if match(.lparen) != nil {
            // Empty tuple ()
            if peek().type == .rparen {
                _ = advance()
                return try parsePostfix(TupleLiteral(elements: []))
            }
            let firstExpr = try parseExpression()
            // Check if this is a tuple (has comma) or just grouped expression
            if match(.comma) != nil {
                var elements: [ASTNode] = [firstExpr]
                if peek().type != .rparen {
                    elements.append(try parseExpression())
                    while match(.comma) != nil {
                        elements.append(try parseExpression())
                    }
                }
                _ = try expect(.rparen)
                return try parsePostfix(TupleLiteral(elements: elements))
            }
            _ = try expect(.rparen)
            return try parsePostfix(firstExpr)
        }

        // Array literal
        if match(.lbracket) != nil {
            var elements: [ASTNode] = []
            if peek().type != .rbracket {
                elements.append(try parseExpression())
                while match(.comma) != nil {
                    elements.append(try parseExpression())
                }
            }
            _ = try expect(.rbracket)
            return try parsePostfix(ArrayLiteral(elements: elements))
        }

        // Dictionary literal: {key: value, ...}
        if peek().type == .lbrace {
            // Look ahead to see if this is a dict literal (has colon after first expr)
            let startPos = pos
            _ = advance() // consume {
            if peek().type == .rbrace {
                _ = advance() // empty dict {}
                return try parsePostfix(DictLiteral(pairs: []))
            }
            // Parse first key
            let firstKey = try parseExpression()
            if peek().type == .colon {
                // This is a dictionary literal
                _ = advance() // consume :
                let firstValue = try parseExpression()
                var pairs: [(key: ASTNode, value: ASTNode)] = [(firstKey, firstValue)]
                while match(.comma) != nil {
                    let key = try parseExpression()
                    _ = try expect(.colon)
                    let value = try parseExpression()
                    pairs.append((key, value))
                }
                _ = try expect(.rbrace)
                return try parsePostfix(DictLiteral(pairs: pairs))
            } else {
                // Not a dict literal, restore position (shouldn't happen in well-formed code)
                pos = startPos
            }
        }

        if let token = match(.number) {
            let numType = token.numericType.flatMap { NumericType(rawValue: $0) }
            return Literal(value: token.value, numericType: numType)
        }
        if let token = match(.interpString) {
            if let rawParts = token.value as? [(isVar: Bool, text: String)] {
                let parts = rawParts.map { part -> StringInterpPart in
                    part.isVar ? .variable(part.text) : .literal(part.text)
                }
                return StringInterpolation(parts: parts)
            }
        }
        if let token = match(.string) {
            return Literal(value: token.value)
        }
        if match(.true) != nil {
            return Literal(value: true)
        }
        if match(.false) != nil {
            return Literal(value: false)
        }
        if match(.nil) != nil {
            return Literal(value: nil)
        }

        if match(.input) != nil {
            _ = try expect(.action)
            _ = try expect(.grab)
            _ = try expect(.lparen)
            let prompt = try parseExpression()
            _ = try expect(.rparen)
            return InputExpr(prompt: prompt)
        }

        if match(.invoke) != nil {
            _ = try expect(.lbrace)
            let nameToken = try expect(.identifier)
            let name = (nameToken.value as? String) ?? ""
            _ = try expect(.rbrace)
            _ = try expect(.action)
            _ = try expect(.with)
            _ = try expect(.lparen)
            var args: [ASTNode] = []
            if peek().type != .rparen {
                args.append(try parseExpression())
                while match(.comma) != nil {
                    args.append(try parseExpression())
                }
            }
            _ = try expect(.rparen)
            return FuncCall(name: name, args: args)
        }

        if let token = match(.identifier) {
            let name = (token.value as? String) ?? ""
            // Check for error tokens from the lexer
            if name.hasPrefix("Unknown ") || name.hasPrefix("Invalid ") || name.hasPrefix("Unexpected ") {
                throw ParserError.unrecognizedStatement(token: name, line: token.line)
            }
            return try parsePostfix(VarRef(name: name))
        }

        let token = peek()
        let tokenText = token.value.map { "\($0)" } ?? "\(token.type)"
        throw ParserError.unexpectedToken(expected: .identifier, got: token.type, gotValue: tokenText, line: token.line)
    }

    private func parsePostfix(_ expr: ASTNode) throws -> ASTNode {
        var result = expr
        // Handle array indexing: arr[index]
        while match(.lbracket) != nil {
            let index = try parseExpression()
            _ = try expect(.rbracket)
            result = IndexAccess(array: result, index: index)
        }
        return result
    }

    private func parseInlineExpr(_ text: String) throws -> ASTNode {
        let lexer = Lexer(source: text)
        let tokens = lexer.tokenize()
        let parser = Parser(tokens: tokens)
        let expr = try parser.parseExpression()
        // Ensure all tokens were consumed
        if parser.peek().type != .eof {
            let leftover = parser.peek()
            let tokenText = leftover.value.map { "\($0)" } ?? "\(leftover.type)"
            if tokenText.hasPrefix("Unknown ") || tokenText.hasPrefix("Invalid ") {
                throw ParserError.unrecognizedStatement(token: tokenText, line: leftover.line)
            }
            throw ParserError.unexpectedToken(expected: .eof, got: leftover.type, gotValue: tokenText, line: leftover.line)
        }
        return expr
    }
}

public enum ParserError: Error, CustomStringConvertible {
    case unexpectedToken(expected: TokenType, got: TokenType, gotValue: String, line: Int)
    case invalidFunctionSignature(String)
    case unrecognizedStatement(token: String, line: Int)

    public var description: String {
        switch self {
        case .unexpectedToken(let expected, _, let gotValue, let line):
            if gotValue.hasPrefix("Unknown ") || gotValue.hasPrefix("Unexpected ") {
                return "\(gotValue) at line \(line)"
            }
            return "Expected \(expected), got '\(gotValue)' at line \(line)"
        case .invalidFunctionSignature(let sig):
            return "Invalid function signature: \(sig)"
        case .unrecognizedStatement(let token, let line):
            if token.hasPrefix("Unknown ") || token.hasPrefix("Invalid ") || token.hasPrefix("Unexpected ") {
                return "\(token) at line \(line)"
            }
            return "Unrecognized statement '\(token)' at line \(line)"
        }
    }
}

File: jibjab/jjpy/jj/parser.py
"""
JibJab Parser - Parses tokens into AST
Uses shared config from common/jj.json for operator emit values
"""

import re
from typing import List, Optional

from .lexer import Lexer, Token, TokenType, JJ
from .ast import (
    ASTNode, Program, PrintStmt, InputExpr, VarDecl, VarRef, Literal,
    BinaryOp, UnaryOp, LoopStmt, IfStmt, FuncDef, FuncCall, ReturnStmt,
    EnumDef, ArrayLiteral, DictLiteral, TupleLiteral, IndexAccess
)

# Get operator emit values from config
OP = JJ['operators']


class Parser:
    def __init__(self, tokens: List[Token]):
        self.tokens = [t for t in tokens if t.type != TokenType.NEWLINE]
        self.pos = 0

    def peek(self, offset: int = 0) -> Token:
        pos = self.pos + offset
        if pos < len(self.tokens):
            return self.tokens[pos]
        return Token(TokenType.EOF, None, 0, 0)

    def advance(self) -> Token:
        token = self.peek()
        self.pos += 1
        return token

    def match(self, *types: TokenType) -> Optional[Token]:
        if self.peek().type in types:
            return self.advance()
        return None

    def expect(self, type: TokenType) -> Token:
        if self.peek().type == type:
            return self.advance()
        token = self.peek()
        got_value = str(token.value) if token.value is not None else self.token_symbol(token.type)
        if isinstance(got_value, str) and (got_value.startswith("Unknown ") or got_value.startswith("Unexpected ")):
            raise SyntaxError(f"{got_value} at line {token.line}")
        raise SyntaxError(f"Expected {type.name.lower()}, got '{got_value}' at line {token.line}")

    @staticmethod
    def token_symbol(token_type: TokenType) -> str:
        symbols = JJ.get('tokenSymbols', {})
        key = token_type.name.lower()
        if key in symbols:
            return symbols[key]
        # Operator symbols from config
        op_map = {
            'ADD': 'add', 'SUB': 'sub', 'MUL': 'mul', 'DIV': 'div', 'MOD': 'mod',
            'EQ': 'eq', 'NEQ': 'neq', 'LT': 'lt', 'LTE': 'lte', 'GT': 'gt',
            'GTE': 'gte', 'AND': 'and', 'OR': 'or', 'NOT': 'not'
        }
        name = token_type.name
        if name in op_map:
            return JJ['operators'][op_map[name]]['symbol']
        if token_type == TokenType.COMMA:
            return ','
        if token_type == TokenType.EOF:
            return 'end of file'
        if token_type == TokenType.BLOCK_END:
            return JJ['blocks']['end']
        return key

    def parse(self) -> Program:
        statements = []
        while self.peek().type != TokenType.EOF:
            stmt = self.parse_statement()
            if stmt:
                statements.append(stmt)
            else:
                bad = self.advance()
                token_text = str(bad.value) if bad.value is not None else str(bad.type.name.lower())
                if token_text.startswith(("Unknown ", "Invalid ", "Unexpected ")):
                    raise SyntaxError(f"{token_text} at line {bad.line}")
                raise SyntaxError(f"Unrecognized statement '{token_text}' at line {bad.line}")
        return Program(statements)

    def parse_statement(self) -> Optional[ASTNode]:
        if self.peek().type == TokenType.PRINT:
            return self.parse_print()
        if self.peek().type == TokenType.SNAG:
            return self.parse_var_decl()
        if self.peek().type == TokenType.LOOP:
            return self.parse_loop()
        if self.peek().type == TokenType.WHEN:
            return self.parse_if()
        if self.peek().type == TokenType.MORPH:
            return self.parse_func_def()
        if self.peek().type == TokenType.YEET:
            return self.parse_return()
        if self.peek().type == TokenType.ENUM:
            return self.parse_enum_def()
        return None

    def parse_print(self) -> PrintStmt:
        self.advance()  # PRINT
        self.expect(TokenType.ACTION)
        self.expect(TokenType.EMIT)
        self.expect(TokenType.LPAREN)
        expr = self.parse_expression()
        self.expect(TokenType.RPAREN)
        return PrintStmt(expr)

    def parse_var_decl(self) -> VarDecl:
        self.advance()  # SNAG
        self.expect(TokenType.LBRACE)
        name = self.expect(TokenType.IDENTIFIER).value
        self.expect(TokenType.RBRACE)
        self.expect(TokenType.ACTION)
        self.expect(TokenType.VAL)
        self.expect(TokenType.LPAREN)
        value = self.parse_expression()
        self.expect(TokenType.RPAREN)
        return VarDecl(name, value)

    def parse_loop(self) -> LoopStmt:
        token = self.advance()  # LOOP with value
        loop_spec = token.value
        body = self.parse_block()

        # Parse loop specification
        if '..' in loop_spec:
            parts = loop_spec.split(':')
            var = parts[0]
            range_parts = parts[1].split('..')
            start = self.parse_inline_expr(range_parts[0])
            end = self.parse_inline_expr(range_parts[1])
            return LoopStmt(var, start, end, None, None, body)
        elif ':' in loop_spec:
            parts = loop_spec.split(':')
            var = parts[0]
            collection = VarRef(parts[1])
            return LoopStmt(var, None, None, collection, None, body)
        else:
            condition = self.parse_inline_expr(loop_spec)
            return LoopStmt('_', None, None, None, condition, body)

    def parse_if(self) -> IfStmt:
        token = self.advance()  # WHEN with condition
        condition = self.parse_inline_expr(token.value)
        then_body = self.parse_block()
        else_body = None

        if self.peek().type == TokenType.ELSE:
            self.advance()
            else_body = self.parse_block()

        return IfStmt(condition, then_body, else_body)

    def parse_func_def(self) -> FuncDef:
        token = self.advance()  # MORPH with signature
        sig = token.value
        match = re.match(r'(\w+)\(([^)]*)\)', sig)
        name = match.group(1)
        params = [p.strip() for p in match.group(2).split(',') if p.strip()]
        body = self.parse_block()
        return FuncDef(name, params, body)

    def parse_return(self) -> ReturnStmt:
        self.advance()  # YEET
        self.expect(TokenType.LBRACE)
        value = self.parse_expression()
        self.expect(TokenType.RBRACE)
        return ReturnStmt(value)

    def parse_enum_def(self) -> EnumDef:
        self.advance()  # ENUM
        self.expect(TokenType.LBRACE)
        name = self.expect(TokenType.IDENTIFIER).value
        self.expect(TokenType.RBRACE)
        self.expect(TokenType.ACTION)
        self.expect(TokenType.CASES)
        self.expect(TokenType.LPAREN)
        cases = []
        if self.peek().type != TokenType.RPAREN:
            cases.append(self.expect(TokenType.IDENTIFIER).value)
            while self.match(TokenType.COMMA):
                cases.append(self.expect(TokenType.IDENTIFIER).value)
        self.expect(TokenType.RPAREN)
        return EnumDef(name, cases)

    def parse_block(self) -> List[ASTNode]:
        statements = []
        while self.peek().type not in (TokenType.BLOCK_END, TokenType.ELSE, TokenType.EOF):
            stmt = self.parse_statement()
            if stmt:
                statements.append(stmt)
            else:
                bad = self.advance()
                token_text = str(bad.value) if bad.value is not None else str(bad.type.name.lower())
                if token_text.startswith(("Unknown ", "Invalid ", "Unexpected ")):
                    raise SyntaxError(f"{token_text} at line {bad.line}")
                raise SyntaxError(f"Unrecognized statement '{token_text}' at line {bad.line}")
        if self.peek().type == TokenType.BLOCK_END:
            self.advance()
        return statements

    def parse_expression(self) -> ASTNode:
        return self.parse_or()

    def parse_or(self) -> ASTNode:
        left = self.parse_and()
        while self.match(TokenType.OR):
            right = self.parse_and()
            left = BinaryOp(left, OP['or']['emit'], right)
        return left

    def parse_and(self) -> ASTNode:
        left = self.parse_equality()
        while self.match(TokenType.AND):
            right = self.parse_equality()
            left = BinaryOp(left, OP['and']['emit'], right)
        return left

    def parse_equality(self) -> ASTNode:
        left = self.parse_comparison()
        while True:
            if self.match(TokenType.EQ):
                left = BinaryOp(left, OP['eq']['emit'], self.parse_comparison())
            elif self.match(TokenType.NEQ):
                left = BinaryOp(left, OP['neq']['emit'], self.parse_comparison())
            else:
                break
        return left

    def parse_comparison(self) -> ASTNode:
        left = self.parse_additive()
        while True:
            if self.match(TokenType.LTE):
                left = BinaryOp(left, OP['lte']['emit'], self.parse_additive())
            elif self.match(TokenType.LT):
                left = BinaryOp(left, OP['lt']['emit'], self.parse_additive())
            elif self.match(TokenType.GTE):
                left = BinaryOp(left, OP['gte']['emit'], self.parse_additive())
            elif self.match(TokenType.GT):
                left = BinaryOp(left, OP['gt']['emit'], self.parse_additive())
            else:
                break
        return left

    def parse_additive(self) -> ASTNode:
        left = self.parse_multiplicative()
        while True:
            if self.match(TokenType.ADD):
                left = BinaryOp(left, OP['add']['emit'], self.parse_multiplicative())
            elif self.match(TokenType.SUB):
                left = BinaryOp(left, OP['sub']['emit'], self.parse_multiplicative())
            else:
                break
        return left

    def parse_multiplicative(self) -> ASTNode:
        left = self.parse_unary()
        while True:
            if self.match(TokenType.MUL):
                left = BinaryOp(left, OP['mul']['emit'], self.parse_unary())
            elif self.match(TokenType.DIV):
                left = BinaryOp(left, OP['div']['emit'], self.parse_unary())
            elif self.match(TokenType.MOD):
                left = BinaryOp(left, OP['mod']['emit'], self.parse_unary())
            else:
                break
        return left

    def parse_unary(self) -> ASTNode:
        if self.match(TokenType.NOT):
            return UnaryOp(OP['not']['emit'], self.parse_unary())
        return self.parse_primary()

    def parse_primary(self) -> ASTNode:
        # Parentheses: either grouped expression or tuple
        if self.match(TokenType.LPAREN):
            # Empty tuple ()
            if self.peek().type == TokenType.RPAREN:
                self.advance()
                return self.parse_postfix(TupleLiteral([]))
            first_expr = self.parse_expression()
            # Check if this is a tuple (has comma) or just grouped expression
            if self.match(TokenType.COMMA):
                elements = [first_expr]
                if self.peek().type != TokenType.RPAREN:
                    elements.append(self.parse_expression())
                    while self.match(TokenType.COMMA):
                        elements.append(self.parse_expression())
                self.expect(TokenType.RPAREN)
                return self.parse_postfix(TupleLiteral(elements))
            self.expect(TokenType.RPAREN)
            return self.parse_postfix(first_expr)

        # Array literal
        if self.match(TokenType.LBRACKET):
            elements = []
            if self.peek().type != TokenType.RBRACKET:
                elements.append(self.parse_expression())
                while self.match(TokenType.COMMA):
                    elements.append(self.parse_expression())
            self.expect(TokenType.RBRACKET)
            return self.parse_postfix(ArrayLiteral(elements))

        # Dictionary literal: {key: value, ...}
        if self.peek().type == TokenType.LBRACE:
            start_pos = self.pos
            self.advance()  # consume {
            if self.peek().type == TokenType.RBRACE:
                self.advance()  # empty dict {}
                return self.parse_postfix(DictLiteral([]))
            # Parse first key
            first_key = self.parse_expression()
            if self.peek().type == TokenType.COLON:
                # This is a dictionary literal
                self.advance()  # consume :
                first_value = self.parse_expression()
                pairs = [(first_key, first_value)]
                while self.match(TokenType.COMMA):
                    key = self.parse_expression()
                    self.expect(TokenType.COLON)
                    value = self.parse_expression()
                    pairs.append((key, value))
                self.expect(TokenType.RBRACE)
                return self.parse_postfix(DictLiteral(pairs))
            else:
                # Not a dict literal, restore position
                self.pos = start_pos

        if token := self.match(TokenType.NUMBER):
            return Literal(token.value)
        if token := self.match(TokenType.STRING):
            return Literal(token.value)
        if self.match(TokenType.TRUE):
            return Literal(True)
        if self.match(TokenType.FALSE):
            return Literal(False)
        if self.match(TokenType.NIL):
            return Literal(None)

        if self.match(TokenType.INPUT):
            self.expect(TokenType.ACTION)
            self.expect(TokenType.GRAB)
            self.expect(TokenType.LPAREN)
            prompt = self.parse_expression()
            self.expect(TokenType.RPAREN)
            return InputExpr(prompt)

        if self.match(TokenType.INVOKE):
            self.expect(TokenType.LBRACE)
            name = self.expect(TokenType.IDENTIFIER).value
            self.expect(TokenType.RBRACE)
            self.expect(TokenType.ACTION)
            self.expect(TokenType.WITH)
            self.expect(TokenType.LPAREN)
            args = []
            if self.peek().type != TokenType.RPAREN:
                args.append(self.parse_expression())
                while self.match(TokenType.COMMA):
                    args.append(self.parse_expression())
            self.expect(TokenType.RPAREN)
            return FuncCall(name, args)

        if token := self.match(TokenType.IDENTIFIER):
            name = token.value
            # Check for error tokens from the lexer
            if isinstance(name, str) and (name.startswith("Unknown ") or name.startswith("Invalid ") or name.startswith("Unexpected ")):
                raise SyntaxError(f"{name} at line {token.line}")
            return self.parse_postfix(VarRef(name))

        token = self.peek()
        token_text = str(token.value) if token.value is not None else self.token_symbol(token.type)
        raise SyntaxError(f"Expected identifier, got '{token_text}' at line {token.line}")

    def parse_postfix(self, expr: ASTNode) -> ASTNode:
        """Parse postfix operations like array indexing"""
        result = expr
        while self.match(TokenType.LBRACKET):
            index = self.parse_expression()
            self.expect(TokenType.RBRACKET)
            result = IndexAccess(result, index)
        return result

    def parse_inline_expr(self, text: str) -> ASTNode:
        """Parse an inline expression from WHEN/LOOP conditions"""
        lexer = Lexer(text)
        tokens = lexer.tokenize()
        parser = Parser(tokens)
        expr = parser.parse_expression()
        # Ensure all tokens were consumed
        if parser.peek().type != TokenType.EOF:
            leftover = parser.peek()
            token_text = str(leftover.value) if leftover.value is not None else str(leftover.type.name.lower())
            if token_text.startswith(("Unknown keyword", "Invalid hash", "Unknown operator")):
                raise SyntaxError(f"{token_text} at line {leftover.line}")
            raise SyntaxError(f"Unexpected '{token_text}' in expression at line {leftover.line}")
        return expr
